File: frontend/apps/ios/Pantopus/Core/Networking/Models/Homes/PostcardDTOs.swift
//
//  PostcardDTOs.swift
//  Pantopus
//
//  DTOs for the postcard residency-verification flow in
//  `backend/routes/homeOwnership.js`:
//    - POST /api/homes/:id/request-postcard (line 2452)
//    - POST /api/homes/:id/verify-postcard  (line 2548)
//
//  Live status exposes saved requests and receipt uncertainty, not USPS tracking.
//

import Foundation

/// Postcard metadata returned by `POST /api/homes/:id/request-postcard`.
public struct PostcardInfoDTO: Decodable, Sendable, Hashable {
    public let id: String
    public let requestedAt: String?
    public let expiresAt: String?

    private enum CodingKeys: String, CodingKey {
        case id
        case requestedAt = "requested_at"
        case expiresAt = "expires_at"
    }
}

/// 201 envelope for `POST /api/homes/:id/request-postcard`.
public struct RequestPostcardResponse: Decodable, Sendable, Hashable {
    public let message: String
    public let postcard: PostcardInfoDTO
    public let deliveryUnknown: Bool?

    private enum CodingKeys: String, CodingKey {
        case message, postcard
        case deliveryUnknown = "delivery_unknown"
    }
}

/// Body for `POST /api/homes/:id/verify-postcard`. Mirrors
/// `verifyPostcardSchema` — a 6–8 char alphanumeric code.
public struct VerifyPostcardRequest: Encodable, Sendable, Hashable {
    public let code: String

    public init(code: String) {
        self.code = code
    }
}

/// 200 envelope for `POST /api/homes/:id/verify-postcard`. The
/// `occupancy` row the handler also returns isn't modelled — the screen
/// only needs to know the call succeeded and the resulting status.
public struct VerifyPostcardResponse: Decodable, Sendable, Hashable {
    public let message: String
    /// `'verified' | 'provisional'`.
    public let verificationStatus: String?
    public let challengeWindowEndsAt: String?

    private enum CodingKeys: String, CodingKey {
        case message
        case verificationStatus = "verification_status"
        case challengeWindowEndsAt = "challenge_window_ends_at"
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/homes/PostcardDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.homes

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * DTOs for the postcard ownership-verification flow in
 * `backend/routes/homeOwnership.js`:
 *   - POST /api/homes/:id/request-postcard (line 2452)
 *   - POST /api/homes/:id/verify-postcard  (line 2548)
 *
 * Note: the backend has no postcard *delivery-tracking* surface, so the
 * A12.7 timeline is driven by `PostcardVerificationSampleData`.
 * Field-for-field parity with iOS `PostcardDTOs.swift`.
 */

/** Postcard metadata returned by `POST /api/homes/:id/request-postcard`. */
@JsonClass(generateAdapter = true)
data class PostcardInfoDto(
    val id: String,
    @Json(name = "requested_at") val requestedAt: String? = null,
    @Json(name = "expires_at") val expiresAt: String? = null,
)

/** 201 envelope for `POST /api/homes/:id/request-postcard`. */
@JsonClass(generateAdapter = true)
data class RequestPostcardResponse(
    val message: String,
    val postcard: PostcardInfoDto,
)

/**
 * Body for `POST /api/homes/:id/verify-postcard`. Mirrors
 * `verifyPostcardSchema` — a 6–8 char alphanumeric code.
 */
@JsonClass(generateAdapter = true)
data class VerifyPostcardRequest(
    val code: String,
)

/**
 * 200 envelope for `POST /api/homes/:id/verify-postcard`. The
 * `occupancy` row the handler also returns isn't modelled — the screen
 * only needs to know the call succeeded.
 */
@JsonClass(generateAdapter = true)
data class VerifyPostcardResponse(
    val message: String,
    @Json(name = "verification_status") val verificationStatus: String? = null,
    @Json(name = "challenge_window_ends_at") val challengeWindowEndsAt: String? = null,
)
